ledger/sandbox/src/main/scala/com/digitalasset/platform/index/LedgerBackedIndexService.scala

```diff
@@ -124,47 +124,52 @@ abstract class LedgerBackedIndexService(
             .toList
       }
 
-  private class OffsetConverter {
+  // Returns a function that memoizes the current end
+  // Can be used directly or shared throughout a request processing
+  private def convertOffset: LedgerOffset => Source[Offset, NotUsed] = {
     lazy val currentEnd: Offset = ledger.ledgerEnd
-
-    def toAbsolute(offset: LedgerOffset): Source[Offset, NotUsed] = offset match {
-      case LedgerOffset.LedgerBegin => Source.single(Offset.begin)
-      case LedgerOffset.LedgerEnd => Source.single(currentEnd)
-      case LedgerOffset.Absolute(offset) =>
-        ApiOffset.fromString(offset).fold(Source.failed, off => Source.single(off))
-    }
+    domainOffset: LedgerOffset =>
+      domainOffset match {
+        case LedgerOffset.LedgerBegin => Source.single(Offset.begin)
+        case LedgerOffset.LedgerEnd => Source.single(currentEnd)
+        case LedgerOffset.Absolute(offset) =>
+          ApiOffset.fromString(offset).fold(Source.failed, off => Source.single(off))
+      }
   }
 
-  private def acceptedTransactions(
+  private def between[A](
       startExclusive: domain.LedgerOffset,
-      endInclusive: Option[domain.LedgerOffset])
-    : Source[(LedgerOffset.Absolute, LedgerEntry.Transaction), NotUsed] = {
-
-    val converter = new OffsetConverter()
-
-    converter.toAbsolute(startExclusive).flatMapConcat { begin =>
+      endInclusive: Option[domain.LedgerOffset],
+  )(f: (Offset, Option[Offset]) => Source[A, NotUsed]): Source[A, NotUsed] = {
+    val convert = convertOffset
+    convert(startExclusive).flatMapConcat { begin =>
       endInclusive
-        .map(converter.toAbsolute(_).map(Some(_)))
+        .map(convert(_).map(Some(_)))
         .getOrElse(Source.single(None))
         .flatMapConcat {
           case Some(`begin`) =>
             Source.empty
-
           case Some(end) if begin > end =>
-            Source.failed(ErrorFactories.invalidArgument(
-              s"End offset ${end.toApiString} is before Begin offset ${begin.toApiString}."))
-
-          case endOpt @ (None | Some(_)) =>
-            ledger
-              .ledgerEntries(Some(begin), endOpt)
-              .collect {
-                case (offset, t: LedgerEntry.Transaction) =>
-                  (toAbsolute(offset), t)
-              }
+            Source.failed(
+              ErrorFactories.invalidArgument(
+                s"End offset ${end.toApiString} is before Begin offset ${begin.toApiString}."))
+          case endOpt: Option[Offset] =>
```

Comment:
`:` doesn't start an ascription here, only a pattern.

```suggestion
          case endOpt =>
```

Reply:
Sorry, but I don't know what 'ascription' means. Can you explain?

Reply:
Ascriptions (exp: t) are static; this syntax implies a dynamic, runtime check.

Reply:
I see. I simply preserved the original, which was `None | Some(_)` and that was motivated [here](https://github.com/digital-asset/daml/pull/4949#discussion_r391514099). Thoughts?